(#507) Change platform_buffer_create()/destroy() interfaces to platform_buffer_init() / deinit() methods.

gapisback opened the pull request:
This commit reworks the interfaces of `platform_buffer_create()` and `platform_buffer_destroy()` to now become `platform_buffer_init()` and `platform_buffer_deinit()` interfaces.

This removes a dependency on `platform_heap_id` or `platform_heap_handle` arg. 

This change does not functionally change anything in these methods. Also, this (indirectly) fixes a minor bug in the create function to deal with the (very rare) case of failing to allocate memory from the heap for `buffer_handle *bh` . Added small unit-test, platform_apis_test, to exercise these changes.

---
NOTE: This is some common infra-code built in a prototype dev branch for supporting different memory handles.

I'm pulling this piece out on its own for integration to /main, so that this ground work can be laid for integrating future dev-efforts (w/ little less pain).

gapisback commented on Makefile:
```diff
@@ -428,6 +428,9 @@ $(BINDIR)/$(UNITDIR)/task_system_test: $(UTIL_SYS)
 
+$(BINDIR)/$(UNITDIR)/platform_apis_test: $(UTIL_SYS)       \
+                                         $(PLATFORM_SYS)
+
```
Minor new unit-test to exercise the changed APIs.

gapisback commented on src/platform_linux/platform.c:
```diff
@@ -56,38 +56,46 @@ void
-                       platform_heap_handle UNUSED_PARAM(heap_handle),
-                       platform_module_id   UNUSED_PARAM(module_id))
+platform_buffer_create(size_t             length,
+                       platform_heap_id   heap_id,
```
This arg sets the stage to allow diff heap-ID handles in future.

gapisback commented on src/platform_linux/platform.c:
```diff
@@ -56,38 +56,46 @@ void
-   buffer_handle *bh = TYPED_MALLOC(platform_get_heap_id(), bh);
+   buffer_handle *bh = TYPED_MALLOC(heap_id, bh);
+   if (bh == NULL) {
+      return bh;
```
Minor bug-fix; otherwise old L99 will seg-fault.

gapisback commented on src/platform_linux/platform.c:
```diff
@@ -103,16 +111,21 @@ platform_buffer_getaddr(const buffer_handle *bh)
    }
 
-   platform_free(platform_get_heap_id(), bh);
+   platform_free(heap_id, *bh_in);
```
Interface changed (for the same reasons as for the `create()` method) so that we can use the right `heap_id` while freeing memory. Otherwise, default `platform_get_heap_id()` may be NULL, and we may end up not freeing the memory with the proper heap-ID.

gapisback commented on src/platform_linux/platform.c:
```diff
@@ -56,38 +56,46 @@ void
-      if (platform_use_hugetlb) {
-         flags |= MAP_HUGETLB;
-      }
+   int prot  = PROT_READ | PROT_WRITE;
```
Blocks of code are indented to the right now as we've lost L 66 `if()` condition. Function will return early on L73 if `bh == NULL`.

gapisback commented on src/platform_linux/platform.c:
```diff
@@ -103,16 +111,21 @@ platform_buffer_getaddr(const buffer_handle *bh)
+ */
 platform_status
-platform_buffer_destroy(buffer_handle *bh)
+platform_buffer_destroy(platform_heap_id heap_id, buffer_handle **bh_in)
```
Changed the interface to receive address of `buffer_handle *`, so that it can be NULL'ed out on exit.
(This is a safer way to avoid callers accidentally accessing `bh_in` after this call.

gapisback commented on src/rc_allocator.c:
```diff
@@ -388,7 +388,7 @@ rc_allocator_init(rc_allocator        *al,
 rc_allocator_deinit(rc_allocator *al)
 {
-   platform_buffer_destroy(al->bh);
+   platform_buffer_destroy(al->heap_id, &al->bh);
```
example of this usage. `al->bh` will be NULL'ed out upon return.

rtjohnso commented on src/platform_linux/platform.h:
```diff
@@ -627,15 +627,15 @@ void
 
 platform_status
-platform_buffer_destroy(buffer_handle *bh);
+platform_buffer_destroy(platform_heap_id heap_id, buffer_handle **bh_in);
```
Can you just add a field to the buffer handle so it remembers its own heapid so you don't need to pass the heapid to the destroy function?

rtjohnso replied:
This comment is obviated by my high level proposal.  I think the high-level proposal will eliminate the need for a heap parameter altogether.

gapisback replied:
Like you observed, elsewhere, this new parameter `heap_id` became unnecessary for these interfaces. I didn't have to add it to the `buffer_handle` struct either, which keeps this change simpler.

rtjohnso reviewed (changes requested):
I would like to propose a different approach:

Change these from `create` and `destroy` functions to `init` and `deinit` functions.

Then change all users to declare `buffer_handle`s instead of `buffer_handle *` fields in their structures.

I believe this will solve the problem you are trying to solve.

I discussed this with Alex and he also thinks this would actually be better than the existing code.

gapisback commented:
HI @rtjohnso - Based on your [suggestion to rework](https://github.com/vmware/splinterdb/pull/508#pullrequestreview-1239454089) this approach, I made the required changes. Uploaded a [new commit, here.](https://github.com/vmware/splinterdb/pull/508/commits/b7a2b22fd926748fe3aba4f92512a68c8e246aa6) 

(I will squash the two commits and will clean up the commit messages once you approve this work, and before I integrate this to `/main`.)

It has certainly simplified the interface and code and solves the dependency on the `heap_id` argument entirely. This works well to pre-stage this change to absorb use of different memory streams in the future. So, thanks!

rtjohnso commented on src/platform_linux/platform.c:
```diff
@@ -56,45 +56,46 @@ void
-            goto error;
-         }
+   if (platform_use_mlock) {
+      int rc = mlock(bh->addr, length);
```
Not you fault but...

I just noticed that this `rc` shadows the `platform_status rc` declared on line 69.

Also, when the `mlock` fails, we return `STATUS_NO_MEMORY`.

Could you change the name of this `rc` so it doesn't shadow the previous, and set a better error code for our return in this case?  Maybe `STATUS_NO_PERMISSION`?

rtjohnso replied:
Oh or maybe `CONST_STATUS(errno)`?

gapisback replied:
Renamed `rc` on L85 to `mlock_rv` and on error, will set `rc` to `CONST_STATUS(errno)`.

rtjohnso commented on src/clockcache.c:
```diff
@@ -1841,11 +1843,12 @@ clockcache_init(clockcache          *cc,   // OUT
-   if (!cc->rc_bh) {
+
+   rc = platform_buffer_init(&cc->rc_bh, refcount_size);
+   if (!SUCCESS(rc)) {
```
The error handling needs to be tightened up.

If either `platform_buffer_init` fails, then we `goto alloc_error`, which does `clockcache_deinit`.  And `clockcache_deinit` unconditionally calls `platform_buffer_deinit` on both buffers.  And `platform_buffer_deinit` unconditionally calls `munmap` on the `buffer_handle`'s `addr`.  But, depending on where we fail, none, one, or both of the buffers may have not been inited, so calling `munmap` could cause an error.

There are two ways to fix this, I think:

1. Modify `clockcache_deinit` to check whether the corresponding pointer is `NULL` before it calls `platform_buffer_deinit`.  So, it would check `cc->data` before calling `platform_buffer_deinit(&cc->bh)` and it would check `cc->refcount` before calling `platform_buffer_deinit(&cc->rc_bh)`.
2. Modify `clockcache_init` to have different error handling targets and do the cleanup on its own, i.e. have `goto alloc_failure1;`, `goto alloc_failure2;`, etc.

My preference is Option 1, but either is fine.

gapisback replied:
@rtjohnso - There are quite a few callers to `platform_buffer_deinit()`, and technically, one has to comb through all instances to make sure that we are not calling this wrongly.

A 3rd approach would be to push the checks to inside of `platform_buffer_deinit()`.

It will try to do the `munmap()` only if `bh->addr` is non-NULL. This will solve the error handling issue from `clockcache_deinit()` and also protect all other instances where `platform_buffer_deinit()` is being called.

What are your thoughts on this approach (before I implement it).

gapisback replied:
(And, ooops, ... sorry, I introduced this error handling issue when implementing the change to introduce `init() / deinit()` interface, without looking too hard at error handling. Sorry!)

gapisback replied:
@rtjohnso : Re: your comment that "_Before the buffer_handle is inited, addr may be garbage, so you cannot safely deinit based on the value of addr._"

In the two places we are talking about, both `rc_allocator_init()` and `clockcache_init()` do a `ZERO_CONTENTS(al)` and `ZERO_CONTENTS(cc)`.

So, certainly for the error path going through `clockcache_deinit(), we should be sure that `cc->bh->addr` will be NULL even before `init()` happened.

Unless you can spot something else I am overlooking, seems like pushing the NULL-check to inside `platform_buffer_deinit()` should work.

gapisback replied:
Discussed this in PR meeting. Rework this as suggested by @robj. `clockcache_deinit()` will only call `platform_buffer_deinit()` only if corresponding pointers are non-NULL.

Remove the fix as applied in the latter function.

rtjohnso commented on src/clockcache.c:
```diff
@@ -1888,15 +1891,12 @@ clockcache_deinit(clockcache *cc) // IN/OUT
-   }
+   platform_buffer_deinit(&cc->rc_bh);
 
    platform_free(cc->heap_id, cc->entry);
```
Along similar lines, I think all of these need to be wrapped in NULL tests.

gapisback replied:
That would not be needed if we pushed the NULL-check to inside `platform_buffer_deinit()`. No ?

gapisback replied:
I misunderstood your comment. Yes, you are right; the calls to `platform_free()` will now only be done if the ptr-variable is non-NULL. Fixed.

rtjohnso commented on src/rc_allocator.c:
```diff
@@ -441,9 +440,14 @@ rc_allocator_mount(rc_allocator        *al,
-   platform_assert(al->bh != NULL);
-   al->ref_count = platform_buffer_getaddr(al->bh);
+   status             = platform_buffer_init(&al->bh, buffer_size);
+   if (!SUCCESS(status)) {
```
Nice!  Love getting rid of assertions.

rtjohnso commented on src/rc_allocator.c:
```diff
@@ -441,9 +440,14 @@ rc_allocator_mount(rc_allocator        *al,
+   status             = platform_buffer_init(&al->bh, buffer_size);
+   if (!SUCCESS(status)) {
+      platform_mutex_destroy(&al->lock);
+      platform_free(al->heap_id, al->meta_page);
```
Although technically, if you want to undo allocations in the reverse order in which they were done, I believe the `platform_mutex_destroy` should come after the `platform_free`.

gapisback replied:
Yes, right.  Changed it accordingly.

rtjohnso commented on tests/functional/test_functionality.c:
```diff
@@ -18,9 +18,10 @@
 
 void
-destroy_test_splinter_shadow_array(test_splinter_shadow_array *sharr)
+destroy_test_splinter_shadow_array(platform_heap_id            hid,
```
I think you can get rid of the `hid` argument.

gapisback replied:
Done. Deleted.

rtjohnso commented on tests/functional/test_splinter_shadow.c:
```diff
@@ -100,11 +100,12 @@ test_splinter_shadow_create(test_splinter_shadow_tree **tree,
                             platform_heap_id            hid,
```
I wish `test_splinter_shadow_create` were also an `...init`.

But that is definitely outside the scope of this PR, so not your problem!

rtjohnso replied:
Also, `hh` is now unused, so delete it.

rtjohnso commented on tests/functional/test_splinter_shadow.c:
```diff
@@ -323,25 +319,25 @@ test_splinter_shadow_iterate_tree(AvlTreeLinks               *root,
 test_splinter_build_shadow_array(test_splinter_shadow_tree  *tree,
                                  test_splinter_shadow_array *shadow_array,
-                                 platform_heap_handle        hh)
+                                 platform_heap_handle        hid)
```
Looks like `hid` is unused now.

gapisback replied:
Took it out. I did some more cleanup to percolate up the stack, cleaning up fn-args like `hh', `hid` that are no longer needed. Some might have slipped thru ... I'll make another pass once this PR checks-in.

rtjohnso commented on src/clockcache.h:
```diff
@@ -116,14 +116,14 @@ struct clockcache {
+   buffer_handle        bh;   // actual memory for pages
    char                *data; // convenience pointer for bh
    platform_log_handle *logfile;
    platform_heap_handle heap_handle;
```
The `heap_handle` field appears to be dead after these changes.

gapisback replied:
Yes; you are right. Took it out.

rtjohnso commented on src/rc_allocator.h:
```diff
@@ -58,7 +58,7 @@ typedef struct rc_allocator_stats {
    allocator               super;
    allocator_config       *cfg;
-   buffer_handle          *bh;
+   buffer_handle           bh;
```
The `heap_handle` field of this structure is now dead, I think.

gapisback replied:
Took it out. And purged unnecessary `heap_handle *` variables and args in other code where `rc_allocator_init()` was being called.

rtjohnso reviewed (changes requested):
Awesome.

There are two main things I'm requesting:

- Clean up the error handling in `clockcache_init` (see comments).
- Audit the code you've changed to see if any additional `heap_id` or `heap_handle` parameters or structure fields can be eliminated.  I think I flagged them all in comments, but it would be good to double-check.

Thank you!

rtjohnso commented:
Before the buffer_handle is inited, addr may be garbage, so you cannot
safely deinit based on the value of addr.

Best,
Rob

Message ID: ***@***.***>
>

rtjohnso commented:
That's pretty un-idiomatic of how `_init` functions usually work and makes
some weird usage rules.

What if I want to use a buffer_handle somewhere else?  What if I want one
on the stack?  Now I have to know that I need to zero it before I can init
it?  Very error prone.


On Sun, Jan 8, 2023 at 9:44 PM gapisback ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/clockcache.c
> <https://github.com/vmware/splinterdb/pull/508#discussion_r1064297981>:
>
> > @@ -1841,11 +1843,12 @@ clockcache_init(clockcache          *cc,   // OUT
>
>     /* Entry per-thread ref counts */
>     size_t refcount_size = cc->cfg->page_capacity * CC_RC_WIDTH * sizeof(uint8);
> -   cc->rc_bh = platform_buffer_create(refcount_size, cc->heap_handle, mid);
> -   if (!cc->rc_bh) {
> +
> +   rc = platform_buffer_init(&cc->rc_bh, refcount_size);
> +   if (!SUCCESS(rc)) {
>
> @rtjohnso <https://github.com/rtjohnso> : Re: your comment that "*Before
> the buffer_handle is inited, addr may be garbage, so you cannot safely
> deinit based on the value of addr.*"
>
> In the two places we are talking about, both rc_allocator_init() and
> clockcache_init() do a ZERO_CONTENTS(al) and ZERO_CONTENTS(cc).
>
> So, certainly for the error path going through clockcache_deinit(), we
> should be sure that cc->bh->addrwill be NULL even beforeinit()` happened.
>
> Unless you can spot something else I am overlooking, seems like pushing
> the NULL-check to inside platform_buffer_deinit() should work.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/vmware/splinterdb/pull/508#discussion_r1064297981>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA45FRLHOW4SBR2QRAI4NGLWROQR5ANCNFSM6AAAAAATGFT2QU>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

gapisback commented:
@rtjohnso - I implemented much of the cleanup to rip out `heap_handle *` args from many places, but chasing other not-needed `hh` and `hid` will take me some time ( and a big screen ) ... Will follow-up tomorrow.

gapisback commented:
This part, Rob - "_What if I want one on the stack?  Now I have to know that I need to zero it before I can initit?  Very error prone._" .... I'm not so sure, it's that error prone.

I have been trained and have gotten used to a conventional way of coding where every on-stack structure is always `MEMZERO()`-ed; i.e. in our code base `ZERO_CONTENTS()`. For precisely the reason that you want to make sure that all fields are zero'ed out before anything -- including init -- is done with the structure.

I think requiring that on-stack variables, structures are always zero'ed out before use is a stronger discipline to have / enforce.

I can implement it in the ways you are suggesting, but I think then there is more work to chase down all other places where `platform_buffer_init()` is called and ensure that error handling is done right.

We should probably try to talk this through to resolve the right approach to adopt.

gapisback commented on src/clockcache.c:
```diff
@@ -1873,33 +1877,41 @@ clockcache_init(clockcache        *cc,   // OUT
    }
 
-   platform_buffer_deinit(&cc->rc_bh);
+   if (cc->lookup) {
```
I did not bother to arrange these backout codes in `reverse` order of how they are done in `init()` method as each backout fragment is protected under an `if()` check. 

So, although, not in "pure backout form", this ordering should work.

gapisback commented on src/clockcache.c:
```diff
@@ -1873,33 +1877,41 @@ clockcache_init(clockcache        *cc,   // OUT
-   platform_free(cc->heap_id, cc->entry);
-   platform_free(cc->heap_id, cc->lookup);
    platform_buffer_deinit(&cc->bh);
+   platform_buffer_deinit(&cc->rc_bh);
```
I did not bracket this with a corresponding `if (!NULL)` check on the member field as I am relying on the the `bh` struct being zero'ed out on declaration. 

Current code always conforms to this ... Let's talk this point through ...

gapisback commented:
Hi, @rtjohnso & @ajhconway -- I am in the middle of amending the code to correctly handle backout error issues, and to purge  `heap_handle *` args, and fn-params that are no longer needed.

The error handling corrections are partially implemented; see [this commit](https://github.com/vmware/splinterdb/pull/508/commits/43efdcf0fecb6da8b2c0f5718713acfcab60e9c9). 

I paused as this needs some unit-testing and I have provided a sample of such a unit-test here. We need better `errno`-type error reporting back from individual sub-systems [ here, clockcache ], so the unit-test can verify that the backout was triggered at the expected error location. See, e.g., the logic implemented in new test case `test_clockcache_init_fails_to_allocate_lookup_array()`.

I think there are few open items with this work that are best discussed in-person. Let's try to cover these items on Tue's PR meeting.

gapisback commented on src/rc_allocator.c:
```diff
@@ -351,7 +351,7 @@ rc_allocator_init(rc_allocator      *al,
    }
    // To ensure alignment always allocate in multiples of page size.
-   uint32 buffer_size = cfg->extent_capacity * sizeof(uint8);
+   uint64 buffer_size = cfg->extent_capacity * sizeof(uint8);
```
Potential bugfix. `buffer_size` declared as uint32 could overflow, to 0, for very large `extent_capacity`. (Discovered this while trying to build negative test cases in newly added, incomplete, `clockcache_test.c` unit test.

gapisback commented on src/platform_linux/platform.c:
```diff
@@ -82,12 +82,13 @@ platform_buffer_init(buffer_handle *bh, size_t length)
    if (platform_use_mlock) {
-      int rc = mlock(bh->addr, length);
-      if (rc != 0) {
+      int mlock_rv = mlock(bh->addr, length);
```
Rename existing `rc` which shadows the  new instance added at the top of this function (in an earlier commit of this PR cycle).

gapisback commented on src/platform_linux/platform.c:
```diff
@@ -82,12 +82,13 @@ platform_buffer_init(buffer_handle *bh, size_t length)
                             length,
                             strerror(errno));
          munmap(bh->addr, length);
+         rc = CONST_STATUS(errno);
```
As suggested by Rob ... this is is better.

gapisback commented on tests/unit/btree_stress_test.c:
```diff
@@ -174,6 +172,7 @@ CTEST_TEARDOWN(btree_stress)
    clockcache_deinit(&data->cc);
    rc_allocator_deinit(&data->al);
    task_system_destroy(data->hid, &data->ts);
+   platform_heap_destroy(&data->hh);
```
Missing in initial code; not an issue right now ... (Not sure why ASAN-builds are not tripping up ...)

gapisback commented on src/platform_linux/platform.c:
```diff
@@ -112,13 +114,16 @@ platform_buffer_getaddr(const buffer_handle *bh)
-   if (ret) {
-      return CONST_STATUS(errno);
-   }
+   if (bh->addr) {
```
Discuss with @rtjohnso -- is this acceptable to you now?

gapisback commented on src/platform_linux/platform.c:
```diff
@@ -112,13 +114,16 @@ platform_buffer_getaddr(const buffer_handle *bh)
-   bh->addr = NULL;
+      bh->addr = NULL;
+   }
+   bh->length = 0;
```
This and L124 are added so that we can validate the state of resulting output `bh` in unit-tests.

gapisback commented on tests/functional/test_functionality.c:
```diff
@@ -632,19 +630,18 @@ cmp_ptrs(const void *a, const void *b)
-                   uint64               num_inserts,
-                   uint64               correctness_check_frequency,
-                   task_system         *state,
-                   platform_heap_handle hh,
```
Deleting this arg caused clang-format to throw a fit.

gapisback commented on tests/unit/platform_apis_test.c:
```diff
@@ -24,6 +26,22 @@ CTEST_DATA(platform_api)
+      FILE *dev_null = fopen("/dev/null", "w");
+      ASSERT_NOT_NULL(dev_null);
+      platform_set_log_streams(dev_null, dev_null);
+   }
```
This block will be simplified with just call to one function when PR #514 checks-in to `/main`.

gapisback commented on tests/unit/platform_apis_test.c:
```diff
@@ -45,13 +63,39 @@ CTEST2(platform_api, test_platform_buffer_init)
+   size_t length = (1024 * KiB * GiB);
+
+   // This is expected to fail as length is very large
+   rc = platform_buffer_init(&bh, length);
```
From code inspection and manual testing, the length set on L90 is sufficiently big-enough to cause `mmap()` to fail w/ an error.

Should this approach be tightened further? (I think this is sufficient.)

gapisback commented:
@rtjohnso - As per our discussion in the PR review meeting, I reworked the flow in `clockcache_deinit()` and adjusted the logic accordingly in `platform_buffer_deinit()`.

This [latest commit](https://github.com/vmware/splinterdb/pull/508/commits/f422a168ec175e4fd80619b47ebded653d3df327) (squashing 2 prev related-commits) should address the last remaining issue.

rtjohnso commented on src/clockcache.c:
```diff
@@ -1847,9 +1847,13 @@ clockcache_init(clockcache        *cc,   // OUT
    cc->pincount =
       TYPED_ARRAY_ZALLOC(cc->heap_id, cc->pincount, cc->cfg->page_capacity);
+   if (!cc->pincount) {
+      goto alloc_error;
```
I think we decided to abandon that level of detail in the testing, right?

gapisback replied:
yes, you are right. I withdrew the newly developed `clockcache_test` unit-test before merging this to /main.

rtjohnso reviewed (approved):
Thank you!  That cleaned up a lot of stuff.

gapisback commented:
@rtjohnso - About your very last comment, I'm only adding the extra check that in case we are unable to allocate memory to `cc->pincount()` via `TYPED_ARRAY_ZALLOC()`, that we bail, and go to error-backout target. (This is not test code; this is mainline library code.)

This was also (already) being done in `/main` here, further below, and also in other places in this `clockcache_init()` function which is allocation memory using `TYPED_ARRAY_ZALLOC()` or variants.

```
1860    cc->batch_busy =
1861       TYPED_ARRAY_ZALLOC(cc->heap_id,
1862                          cc->batch_busy,
1863                          cc->cfg->page_capacity / CC_ENTRIES_PER_BATCH);
1864    if (!cc->batch_busy) {
1865       goto alloc_error;
1866    }
```
